lib/active_merchant/billing/gateways/cashnet.rb

```diff
@@ -110,9 +110,34 @@ def add_creditcard(post, creditcard)
         post[:lname]           = creditcard.last_name
       end
 
-      def add_invoice(post, options)
+      def add_invoice(post, money, options)
         post[:order_number] = options[:order_id] if options[:order_id].present?
-        post[:itemcode] = (options[:item_code] || @options[:default_item_code])
+
+        if options[:item_codes].present?
+          item_codes = options[:item_codes].select { |key, _| key.to_s.start_with?('item_code') }
+
+          item_codes.each do |key, value|
+            index_array = key.to_s.split('item_code')
+
+            if index_array == []
+              post[:itemcode] = value
+            else
+              number = index_array[1]
+              post_key = 'itemcode' + number
+
+              post[post_key] = value
+            end
+          end
```

Comment:
@dsmcclain I've refactored as we discussed yesterday. However, the move from snake case `item_code#{number}` to `itemcode#{number}` feels extremely fragile and clunky. any advice for how to improve this within ruby?

Reply:
Yeah, I feel ya. I do think it is still the right idea to preserve the snake case approach, because part of the promise of Active Merchant is to provide a more-or-less standardized API for merchants that abstracts away the idiosyncrasies of each gateway (in this case, wacky query params).

Ruby provides a `transform_keys` method ([documentation here](https://ruby-doc.org/core-2.6/Hash.html#transform_keys-method)), which I think you could combine with `gsub` to achieve this with a lot less work:
```
if options[:item_codes].present?
  codes_and_amounts = options[:item_codes].transform_keys { |key| key.to_s.gsub('_', '') } # so long, snake case!
  codes_and_amounts.each do |key, value|
    post[key] = value if key.start_with?('itemcode', 'amount') # ignore anything that isn't an item code or amount
  end
else
  ...do other stuff
end
```

One potentially useful side effect of this approach is that if a merchant _did_ end up passing keys like `'itemcode2'` and `'amount2'` it would still go through.

 If this approach still feels fragile to you, it could be nice to add a test case or two showing that passing in an `item_codes` hash with weird keys just causes them to get dropped instead of blowing everything up.

Reply:
@dsmcclain amazing, of course, why split it out when all we need to do is remove the underscore!!! Classic "Why didn't I think of that" moment 😂 

Yeah, was thinking some test cases for this would be useful once I got the transform finalized. Will implement now!